numpy/core/src/multiarray/ctors.c

```diff
@@ -3711,14 +3725,15 @@ PyArray_FromIter(PyObject *obj, PyArray_Descr *dtype, npy_intp count)
     for (i = 0; (i < count || count == -1) &&
              (value = PyIter_Next(iter)); i++) {
         if (i >= elcount) {
+            npy_intp nbytes;
             /*
               Grow PyArray_DATA(ret):
               this is similar for the strategy for PyListObject, but we use
               50% overallocation => 0, 4, 8, 14, 23, 36, 56, 86 ...
             */
             elcount = (i >> 1) + (i < 4 ? 4 : 2) + i;
-            if (elcount <= NPY_MAX_INTP/elsize) {
-                new_data = PyDataMem_RENEW(PyArray_DATA(ret), elcount * elsize);
+            if (!npy_mul_with_overflow_intp(elcount, elsize, &nbytes)) {
```

Comment:
Indeed, but that's pretty late in the pipeline, and it runs a bunch of other tests first